tsl/src/partialize_agg.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <catalog/pg_operator.h>
+#include <miscadmin.h>
+#include <nodes/bitmapset.h>
+#include <nodes/makefuncs.h>
+#include <nodes/nodeFuncs.h>
+#include <optimizer/cost.h>
+#include <optimizer/optimizer.h>
+#include <optimizer/pathnode.h>
+#include <optimizer/paths.h>
+#include <parser/parsetree.h>
+#include <utils/builtins.h>
+#include <utils/lsyscache.h>
+#include <utils/typcache.h>
+#include <utils/fmgroids.h>
+
+#include <planner.h>
+
+#include "nodes/decompress_chunk/decompress_chunk.h"
+#include "partialize_agg.h"
+#include "utils.h"
+#include "debug_assert.h"
+
+#define is_restricted_path(path) list_length(path->parent->baserestrictinfo)
+
+/*
+ * Are we able to optimize the path by applying vectorized aggregation?
+ */
+static bool
+is_vectorizable_agg_path(PlannerInfo *root, AggPath *agg_path, Path *path)
+{
+	Assert(agg_path->aggstrategy == AGG_SORTED || agg_path->aggstrategy == AGG_PLAIN ||
+		   agg_path->aggstrategy == AGG_HASHED);
+
+	/* Having is not supported at the moment */
+	if (root->hasHavingQual)
+		return false;
+
+	/* Only vectorizing within the decompress node is supported so far */
+	bool is_decompress_chunk = ts_is_decompress_chunk_path(path);
+	if (!is_decompress_chunk)
+		return false;
+
+	DecompressChunkPath *decompress_path = (DecompressChunkPath *) path;
+	Assert(decompress_path->custom_path.custom_paths != NIL);
+
+	/*
+	 * It would be great if we could check decompress_path -> have_bulk_decompression_columns
+	 * here (or ensure the column can be bulk decompressed). However, this information is only
+	 * computed when a path is selected and the plan is created.
+	 *
+	 * However, due to the restriction to a particular data type for vectorized aggregates
+	 * (i.e., INT4) we can assume that this column type can be bulk decompressed.
+	 */
```

Comment:
Yeah this is unfortunate... we'll have to do something about this when we move past a prototype. Maybe this should be at the level of Plans as well? Probably we don't need a cost-based decision between whether to vectorize the aggregation or not. We always want to vectorize it when we can. So we don't need to do anything about it at the Path stage. What do you think? Might be better to redo this before merging.

Reply:
To put it in other words, at the Path stage we want a minimal representation of the plan that is needed for cost-based decisions, and vectorized aggregation pushdown doesn't have to be cost-based, so it can go to the Plan stage.

Reply:
@akuzm I rewrote this part of the code. I now check for column type and bulk decompression at the Path stage. 

I decided not to move this into the Plan stage because (1) most of our query-rewrite code works at the Path stage and I would not like to duplicate code, (2) I would like to keep the code for the planning together to make it easier to understand, and (3) we had some discussions of enabling vectorization for time_bucket queries in the future. This is also something that is easier to check at the Path stage. 

Since hypertable_compression_info is already present at this stage, I can use this information without fetching further data from the catalog. Could you have a look at the changes (see the second commit of this PR)?

Reply:
Discussed with @akuzm on Slack that we merge this PR with the additional planner changes.